VerificationResult now has a proper mediaType field.

phillmv opened the pull request:
<!--
Thanks for opening a pull request! Please do not just delete this text.  The three fields below are mandatory.

Please remember to:
- This PR requires an issue. If it is a new feature, the issue should proceed the PR and will have allowed sufficent time for discussions to take place. Please use
issue tags such as "Closes #XYZ" or "Resolves sigstore/repo-name#XYZ".
  [documentation](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- ensure your commits are signed-off, as sigstore uses the [DCO](https://en.wikipedia.org/wiki/Developer_Certificate_of_Origin) using `git commit -s`, or `git commit -s --amend` if you want to amend already existing commits
- lastly, ensure there are no merge commits!

Thank you :)
-->

#### Summary

Closes https://github.com/sigstore/sigstore-go/issues/13

As described in the issue, this PR solves the problem of `VerificationResults.Version` being insufficiently descriptive and incongruous with the other json document types that are handled in the broader Sigstore ecosystem.

This PR:
- Changes it from `Version int` to `MediaType string`
- Updates the default value to `application/vnd.dev.sigstore.verificationresult+json;version=0.1`
- And for good measure, adds the option to use `WithoutArtifactUnsafe()` in the `sigstore-go` binary, which I used to test this feature.

<!--
 Explain the **motivation** for making this change. What existing problem does the pull request solve? How can reviewers test this PR?
-->

#### Release Note

- Changed VerificationResult.Version to VerificationResult.MediaType

<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to an administrator running private sigstore instances (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense.

-->

#### Documentation

We haven't created a release for this yet, so I don't think any additional docs changes are necessary beyond what will be generated when we do create a release for this library.

<!--

Does this change require an update to documentation? How will users implement your new feature?

Please reference a PR within https://docs.sigstore.dev

-->

cpanato reviewed (dismissed):
Cool

asraa commented:
Just fixed lint!

asraa commented:
@haydentherapper want to do a quick skim? this is mostly stolen from rekor's verification/cosign's verification except that it works on the protobuf log entry.

haydentherapper commented on pkg/tlog/verify.go:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	// Extract the SET from the tlog entry
+	if entry.GetInclusionPromise() == nil {
```
this field makes me happy, excellent naming!

asraa replied:
it's protobuf generated code!

haydentherapper reviewed (commented):
mostly just some nits

asraa commented on pkg/tlog/verify_test.go:
```diff
@@ -0,0 +1,167 @@
+)
+
+const (
+	encodedBody = "ayJhcGlWZXJzaW9uIjoiMC4wLjEiLCJraW5kIjoicmVrb3JkIiwic3BlYyI6eyJkYXRhIjp7Imhhc2giOnsiYWxnb3JpdGhtIjoic2hhMjU2IiwidmFsdWUiOiJlY2RjNTUzNmY3M2JkYWU4ODE2ZjBlYTQwNzI2ZWY1ZTliODEwZDkxNDQ5MzA3NTkwM2JiOTA2MjNkOTdiMWQ4In19LCJzaWduYXR1cmUiOnsiY29udGVudCI6Ik1FWUNJUUQvUGRQUW1LV0MxKzBCTkVkNWdLdlFHcjF4eGwzaWVVZmZ2M2prMXp6Skt3SWhBTEJqM3hmQXlXeGx6NGpwb0lFSVYxVWZLOXZua1VVT1NvZVp4QlpQSEtQQyIsImZvcm1 ...
```
oh, it didn't need to be, for clarity i'll just put in an explicit "dummy" value. SET verification doesn't do any unmarshalling of this field.

asraa commented:
Updated with more comments and tests

phillmv committed:
VerificationResult now has a proper mediaType field.

Signed-off-by: Phill MV <phillmv@github.com>

phillmv committed:
Linting.

I actually find the switch statement here MORE confusing than the if else,
so I have taken the liberty to disable gocritic for this particular if.

Signed-off-by: Phill MV <phillmv@github.com>

kommendorkapten commented on pkg/verify/signed_entity.go:
```diff
@@ -174,7 +174,7 @@ type TimestampVerificationResult struct {
 func NewVerificationResult() *VerificationResult {
 	return &VerificationResult{
-		Version: 20230823,
+		MediaType: "application/vnd.dev.sigstore.verificationresult+json;version=0.1",
```
Nit but we should define the media type as a `const` so it's easier to work with.

phillmv committed:
Moved VerificationResult mediaType value to a const.

Signed-off-by: Phill MV <phillmv@github.com>

steiza reviewed (approved):
Looks good to me!
